common/src/main/java/io/netty/util/concurrent/FastThreadLocal.java

```diff
@@ -123,10 +126,29 @@ private static void removeFromVariablesToRemove(
         variablesToRemove.remove(variable);
     }
 
+    // visible for tests
+    static boolean hasOnRemoval(Class<? extends FastThreadLocal> origClass) {
+        for (Class<?> clazz = origClass; clazz != FastThreadLocal.class; clazz = clazz.getSuperclass()) {
+            try {
+                // Can't use getMethod since onRemoval is protected
```

Comment:
didn't know that :(

Reply:
@ejona86 couldn't you use the following ?

```java
static boolean hasOnRemoval(Class<? extends FastThreadLocal> origClass) {
    try {
        origClass.getMethod("onRemoval").getDeclaringClass() != FastThreadLocal.class
    } catch (Throwable cause) {
        // logging.....
        return true;
    }
}

```

Reply:
actually seems like you tried something like this... Can you explain why this not works ?

Reply:
@ejona86 I guess what you could also do is loop through all the classes as you did and call:

`getDeclaredMethods()` on each class and check if `onRemoval` exists there. This should work  with protected as well. See https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getDeclaredMethods()

Reply:
`getMethod()` does not return protected methods, so it throws NoSuchMethodException virtually all the time.

From getMethod's docs:
> Returns a Method object that reflects the specified _public_ member method...

getDeclaredMethod does return protected methods, but it doesn't recurse to the superclass when the method is not found, so the looping was required.

Reply:
> I guess what you could also do is loop through all the classes as you did and call: getDeclaredMethods() on each class and check if onRemoval exists there.

Isn't that what I did?

Reply:
@ejona86 nope... This is what I would do:

```
    static boolean hasOnRemoval(Class<? extends FastThreadLocal> origClass) {
        for (Class<?> clazz = origClass; clazz != FastThreadLocal.class; clazz = clazz.getSuperclass()) {
            try {
                for (Method method: clazz.getDeclaredMethods()) {
                    if ("onRemoval".equals(method.getName())) {
                        Class[] paramTypes = method.getParameterTypes();
                        if (paramTypes.length == 1 && paramTypes[0] == Object.class) {
                            return true;
                        }
                    }
                }
            } catch (Throwable t) {
                logger.info("Failed to determine if onRemoval is overridden", t);
                return true;
            }
        }
        return false;
    }
```

This will not need to cause any exceptions

Reply:
I will look more closely tomorrow but I think this will interact poorly with a security manager and so unlikely to be a solution for us. We require the security manager, run our tests with one, and are going to be unwilling to grant Netty the `accessDeclaredMembers` runtime permission (I think that is what will be needed here).

Reply:
@jasontedor thanks a lot...

Reply:
could we use a marker interface?

Reply:
@jasontedor I am correct that the problem you mentioned is true for my suggested code but also for the one that is used in this PR, correct ?

Reply:
Tagging interface is a bit hard, because existing users of the class won't use it. The reason the reflection is necessary is to continue supporting the current API.

We could create a NoRemovalNotificationFastThreadLocal (NRNFTL; "pending marketing approval") that overrides onRemoval and makes the method `final`. Then FastThreadLocal could see if the current instance extends NRNFTL. That does mean most existing users of FastThreadLocal would need to be changed, which I think does make it a less-nice change. The in-tree ones could be changed to it easily, though.

The current API is unfortunate, because `onRemoval()` usage is the uncommon case. There's only three in-tree users: `PooledByteBufAllocator`, `Recycler`, and epoll's `NativeDatagramPacketArray`. PooledByteBufAllocator could avoid `get()ing` from the thread local when `useCacheForAllThreads == true`. Recycler could be disabled by default, or replaced (#8052). The `NativeDatagramPacketArray` code looks like it may only be called from the event loop.

Reply:
@ejona86 I agree... that said we could also mark `onRemoval` as `@Deprecated` and only handle it from the `FastLocalThread` (we did the same before which I "fixed").  I am somewhat sure we could remove the `onRemoval` usage in our internal use-cases in many if not in all cases. 

Only partial handle onRemoval is a bit unfortunate but maybe the best bet.

WDYT ?

Also /cc @jasontedor @johnou